### PR TITLE
Initial end-of-puzzle achievement checkboxes (behind experiment flag)

### PR DESCRIPTION
**Pull request by joshlory**
To view, add `enableExperiments=gamification` to any local Code Studio url.

These don't match Mike's mocks 100% yet, I've only added the initial checkboxes for achievements.  No tests yet because we plan to iterate a lot on the experiments next sprint.

Completing a puzzle without hints, and with the ideal number of blocks:

screen shot 2017-02-09 at 5 32 14 pm

Completing a puzzle with hints, and too many blocks:

screen shot 2017-02-09 at 5 27 55 pm

### CODE REVIEW

File: apps/src/feedback.js [hunk 1]
```diff
@@ -188,6 +192,23 @@ FeedbackUtils.prototype.displayFeedback = function (options, requiredBlocks,
...
+    ReactDOM.render(
+      <AchievementDialog
+        puzzleNumber={options.level.puzzle_number || 0}
+        idealBlocks={this.studioApp_.IDEAL_BLOCK_NUM}
+        actualBlocks={this.getNumCountableBlocks()}
+        hintsUsed={hintsUsed}
+        assetUrl={this.studioApp_.assetUrl}
+        onContinue={options.onContinue}
+      />, container);
+    document.body.appendChild(container);
```
Comment (Hamms) [inline apps/src/feedback.js]: nit: I'd stick the container in the body before rendering React into it; there are a few components in our system that prefer that order.

File: apps/src/templates/BaseDialog.jsx [hunk 2]
```diff
@@ -77,22 +79,45 @@ var BaseDialog = React.createClass({
     let modalBodyClassNames = "modal-body";
     let modalBackdropClassNames = "modal-backdrop";
 
-    if (this.props.useDeprecatedGlobalStyles) {
+    if (this.props.useUpdatedStyles) {
+      modalBodyClassNames = "";
+      modalBodyStyle = {
+        background: `#fff top center url(${this.props.assetUrl('media/background.png')}) no-repeat`,
```
Comment (Hamms) [inline apps/src/templates/BaseDialog.jsx]: nit: could we call it something more descriptive than "background.png"?

File: dashboard/app/models/user.rb [hunk 3]
```diff
@@ -975,12 +975,8 @@ def self.track_level_progress_sync(user_id:, level_id:, script_id:, new_result:,
...
           ScriptConstants::COURSE4_NAME
         ].include? Script.get_from_cache(script_id).name) &&
-        HintViewRequest.
-          where(user_id: user_id, script_id: script_id, level_id: level_id).
-          empty? &&
-        AuthoredHintViewRequest.
-          where(user_id: user_id, script_id: script_id, level_id: level_id).
-          empty?
+        HintViewRequest.no_hints_used?(user_id, script_id, level_id) &&
+        AuthoredHintViewRequest.no_hints_used?(user_id, script_id, level_id)
```
Comment (Hamms) [inline dashboard/app/models/user.rb:979]: :+1:

File: apps/src/templates/AchievementDialog.jsx [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+            {this.getIcon(!tooManyBlocks)}
+            <span
+              style={[
+                styles.achievement.text,
+                tooManyBlocks && styles.achievement.inactive
+              ]}
+            >
+              {tooManyBlocks ? locale.usingTooManyBlocks(params) :
+                (blockDelta === 0 ? locale.exactNumberOfBlocks(params) :
+                  locale.fewerNumberOfBlocks(params)
```
Comment (Hamms) [inline apps/src/templates/AchievementDialog.jsx]: might be nice to have this logic up by the `footerMessage` definition

Comment (joshlory) [reply apps/src/templates/AchievementDialog.jsx]: Good call.  I tried extracting these into functions in https://github.com/code-dot-org/code-dot-org/pull/13175/commits/4880f8eea116ef1159b2731fcdb365e339f61139 — thoughts?

Comment (Hamms) [reply apps/src/templates/AchievementDialog.jsx]: Beautiful!

**Pull request merged**